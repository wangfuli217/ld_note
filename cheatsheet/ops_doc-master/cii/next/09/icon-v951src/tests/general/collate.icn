procedure main()
   s1 := collate(&cset,&cset)
   s2 := collate(reverse(&cset),reverse(&cset))
   write(image(decollate(s1,0)))
   write(image(decollate(s1,1)))
   write(image(decollate(s2,1)))
   write(image(decollate(s2,0)))
   perm()
end

procedure collate(s1,s2)
  local length, ltemp, rtemp, t
  static llabels, rlabels, clabels, blabels, half
  initial {
    llabels := "abxy"
    rlabels := "cduv"
    blabels := llabels || rlabels
    clabels := "acbdxuyv"
    half := 4
    ltemp := left(&cset,*&cset/2)
    rtemp := right(&cset,*&cset/2)
    clabels := collate(ltemp,rtemp)
    llabels := ltemp
    rlabels := rtemp
    blabels := string(&cset)
    half := *llabels
    }
   if *s1 > *s2 then {
      t := s1[*s2+1:0]
      s1 := s1[1:*s2+1]
      }
   else if *s2 > *s1 then {
      t := s2[*s1+1:0]
      s2 := s2[1:*s1+1]
      }
   else t := ""
  length := *s1
  if length <= half then
    return map(left(clabels,2*length),left(llabels,length) ||
      left(rlabels,length),s1 || s2) || t
  else
    return map(clabels,blabels,left(s1,half) || left(s2,half)) ||
      collate(right(s1,length-half),right(s2,length-half)) || t
end

procedure decollate(s,n)
   static dsize, image, object
   local ssize
   initial {
      image := collate(&cset[2:0],repl(&cset[1],*&cset-1))
      object := string(&cset)
      dsize := *image
      }
   n %:= 2
   ssize := *s
   if ssize + n <= dsize then
      return map(object[1+:(ssize+n)/2],image[(n+1)+:ssize],s)
   else
      return map(object[1+:(dsize-2)/2],image[(n+1)+:dsize-2],
         s[1+:(dsize-2)]) || decollate(s[dsize-1:0],n)
end
procedure perm()
   output := set()
   every 1 to 2 do
      every insert(output,permute("ogram"))
   every write(!sort(output))
end

procedure permute(s)
   local i, x, t
   if s == "" then return ""
   every i := 1 to *s do {
      x := s[i]
      t := s
      t[i] := ""
      suspend x || permute(t)
      }
end
