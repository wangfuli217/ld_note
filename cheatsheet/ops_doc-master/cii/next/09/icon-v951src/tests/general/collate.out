"\xfd\x00\x01\x02\x03\x04\x05\x06\x07\b\t\n\v\f\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\e\x1c\x1d\x1e\x1f !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\d\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0\xb1\xb2\xb3\xb4\xb5\xb6\xb7\xb8\xb9\xba\xbb\xbc\xbd\xbe\xbf\xc0\xc1\xc2\xc3\xc4\xc5\xc6\xc7\xc8\xc9\xca\xcb\xcc\xcd\xce\xcf\xd0\xd1\xd2\xd3\xd4\xd5\xd6\xd7\xd8\xd9\xda\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9\xfa\xfb\xfc\xff\xfe"
"\xfd\x01\x00\x01\x02\x03\x04\x05\x06\x07\b\t\n\v\f\r\x0e\x0f\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1a\e\x1c\x1d\x1e\x1f !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\d\x80\x81\x82\x83\x84\x85\x86\x87\x88\x89\x8a\x8b\x8c\x8d\x8e\x8f\x90\x91\x92\x93\x94\x95\x96\x97\x98\x99\x9a\x9b\x9c\x9d\x9e\x9f\xa0\xa1\xa2\xa3\xa4\xa5\xa6\xa7\xa8\xa9\xaa\xab\xac\xad\xae\xaf\xb0\xb1\xb2\xb3\xb4\xb5\xb6\xb7\xb8\xb9\xba\xbb\xbc\xbd\xbe\xbf\xc0\xc1\xc2\xc3\xc4\xc5\xc6\xc7\xc8\xc9\xca\xcb\xcc\xcd\xce\xcf\xd0\xd1\xd2\xd3\xd4\xd5\xd6\xd7\xd8\xd9\xda\xdb\xdc\xdd\xde\xdf\xe0\xe1\xe2\xe3\xe4\xe5\xe6\xe7\xe8\xe9\xea\xeb\xec\xed\xee\xef\xf0\xf1\xf2\xf3\xf4\xf5\xf6\xf7\xf8\xf9\xfa\xfb\xff\x01"
"\x02\x01\xff\xfe\xfd\xfc\xfb\xfa\xf9\xf8\xf7\xf6\xf5\xf4\xf3\xf2\xf1\xf0\xef\xee\xed\xec\xeb\xea\xe9\xe8\xe7\xe6\xe5\xe4\xe3\xe2\xe1\xe0\xdf\xde\xdd\xdc\xdb\xda\xd9\xd8\xd7\xd6\xd5\xd4\xd3\xd2\xd1\xd0\xcf\xce\xcd\xcc\xcb\xca\xc9\xc8\xc7\xc6\xc5\xc4\xc3\xc2\xc1\xc0\xbf\xbe\xbd\xbc\xbb\xba\xb9\xb8\xb7\xb6\xb5\xb4\xb3\xb2\xb1\xb0\xaf\xae\xad\xac\xab\xaa\xa9\xa8\xa7\xa6\xa5\xa4\xa3\xa2\xa1\xa0\x9f\x9e\x9d\x9c\x9b\x9a\x99\x98\x97\x96\x95\x94\x93\x92\x91\x90\x8f\x8e\x8d\x8c\x8b\x8a\x89\x88\x87\x86\x85\x84\x83\x82\x81\x80\d~}|{zyxwvutsrqponmlkjihgfedcba`_^]\\[ZYXWVUTSRQPONMLKJIHGFEDCBA@?>=<;:9876543210/.-,+*)('&%$#\"! \x1f\x1e\x1d\x1c\e\x1a\x19\x18\x17\x16\x15\x14\x13\x12\x11\x10\x0f\x0e\r\f\v\n\t\b\x07\x06\x05\x04\x00\x01"
"\x02\xff\xfe\xfd\xfc\xfb\xfa\xf9\xf8\xf7\xf6\xf5\xf4\xf3\xf2\xf1\xf0\xef\xee\xed\xec\xeb\xea\xe9\xe8\xe7\xe6\xe5\xe4\xe3\xe2\xe1\xe0\xdf\xde\xdd\xdc\xdb\xda\xd9\xd8\xd7\xd6\xd5\xd4\xd3\xd2\xd1\xd0\xcf\xce\xcd\xcc\xcb\xca\xc9\xc8\xc7\xc6\xc5\xc4\xc3\xc2\xc1\xc0\xbf\xbe\xbd\xbc\xbb\xba\xb9\xb8\xb7\xb6\xb5\xb4\xb3\xb2\xb1\xb0\xaf\xae\xad\xac\xab\xaa\xa9\xa8\xa7\xa6\xa5\xa4\xa3\xa2\xa1\xa0\x9f\x9e\x9d\x9c\x9b\x9a\x99\x98\x97\x96\x95\x94\x93\x92\x91\x90\x8f\x8e\x8d\x8c\x8b\x8a\x89\x88\x87\x86\x85\x84\x83\x82\x81\x80\d~}|{zyxwvutsrqponmlkjihgfedcba`_^]\\[ZYXWVUTSRQPONMLKJIHGFEDCBA@?>=<;:9876543210/.-,+*)('&%$#\"! \x1f\x1e\x1d\x1c\e\x1a\x19\x18\x17\x16\x15\x14\x13\x12\x11\x10\x0f\x0e\r\f\v\n\t\b\x07\x06\x05\x04\x03\x00\x01"
agmor
agmro
agomr
agorm
agrmo
agrom
amgor
amgro
amogr
amorg
amrgo
amrog
aogmr
aogrm
aomgr
aomrg
aorgm
aormg
argmo
argom
armgo
armog
arogm
aromg
gamor
gamro
gaomr
gaorm
garmo
garom
gmaor
gmaro
gmoar
gmora
gmrao
gmroa
goamr
goarm
gomar
gomra
goram
gorma
gramo
graom
grmao
grmoa
groam
groma
magor
magro
maogr
maorg
margo
marog
mgaor
mgaro
mgoar
mgora
mgrao
mgroa
moagr
moarg
mogar
mogra
morag
morga
mrago
mraog
mrgao
mrgoa
mroag
mroga
oagmr
oagrm
oamgr
oamrg
oargm
oarmg
ogamr
ogarm
ogmar
ogmra
ogram
ogrma
omagr
omarg
omgar
omgra
omrag
omrga
oragm
oramg
orgam
orgma
ormag
ormga
ragmo
ragom
ramgo
ramog
raogm
raomg
rgamo
rgaom
rgmao
rgmoa
rgoam
rgoma
rmago
rmaog
rmgao
rmgoa
rmoag
rmoga
roagm
roamg
rogam
rogma
romag
romga
